### PR TITLE
refactor: Replace MSAL with Next Auth

### LINKED ISSUE #125: Authentication in the production environment
We have two URLs available for people to use
- https://web-data-marketplace-prod.radix.equinor.com
- https://data-marketplace.app.radix.equinor.com

The user need to log in to see the application, but this doesn't work properly across browsers and the two URLs.

**To Reproduce**
Go to one of the urls, preferably (`web-data-marketplace`)

It seems like if it works more often with FF than Chrome

**Expected behavior**
A logged in view of the application

**Screenshots**
Skjermbilde 2022-08-10 kl  08 40 21

**Desktop (please complete the following information):**
 - OS: Mainly affects Chrome. Often works in FF. Haven't tested with Edge or Safari yet

**Smartphone (please complete the following information):**

**Additional context**
Logs from [Radix](https://console.radix.equinor.com/applications/data-marketplace)
```
GET request to https://equinor-dev.collibra.com/rest/2.0/navigation/most_viewed failed: 401 ()
[PopularAssetsHandler] HttpError: Request failed with status code 401
    at Function.request (/opt/app/.next/server/chunks/827.js:83:19)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async getPopularAssets (/opt/app/.next/server/pages/api/popular.js:41:29)
    at async PopularAssetsHandler (/opt/app/.next/server/pages/api/popular.js:77:30)
    at async Object.apiResolver (/opt/app/node_modules/next/dist/server/api-utils/node.js:179:9)
    at async NextNodeServer.runApi (/opt/app/node_modules/next/dist/server/next-server.js:381:9)
    at async Object.fn (/opt/app/node_modules/next/dist/server/base-server.js:500:37)
    at async Router.execute (/opt/app/node_modules/next/dist/server/router.js:213:36)
    at async NextNodeServer.run (/opt/app/node_modules/next/dist/server/base-server.js:619:29) {
  body: {
    statusCode: 401,
    titleMessage: 'Unauthorized',
    helpMessage: 'Please check the application logs for more information.',
    userMessage: 'The bearer token is expired',
    errorCode: 'expiredToken'
  },
  headers: {
    date: 'Wed, 10 Aug 2022 06:35:41 GMT',
    'content-type': 'application/json;charset=ISO-8859-1',
    'content-length': '191',
    connection: 'close',
    'set-cookie': [
      'AWSALBTG=KZ9MNZBHnznrruE+Mpe9UTjkSrVi6yPToeV//gPJCR0R6C9fmXN0RozzYw8P4cNgBsNRxuxPJIz3iLOTY25GneKoWWjTgMkge3IltSCM/yE6mtwPfA7R8WH0ggGwv7rQfsNTZRLmslTXVyoGhjl+5qz7yftTe4GBYrwYE2AFHeFBD9G2ZBc=; Expires=Wed, 17 Aug 2022 06:35:41 GMT; Path=/',
      'AWSALBTGCORS=KZ9MNZBHnznrruE+Mpe9UTjkSrVi6yPToeV//gPJCR0R6C9fmXN0RozzYw8P4cNgBsNRxuxPJIz3iLOTY25GneKoWWjTgMkge3IltSCM/yE6mtwPfA7R8WH0ggGwv7rQfsNTZRLmslTXVyoGhjl+5qz7yftTe4GBYrwYE2AFHeFBD9G2ZBc=; Expires=Wed, 17 Aug 2022 06:35:41 GMT; Path=/; SameSite=None; Secure',
      'rememberMe=deleteMe; Path=/; Max-Age=0; Expires=Tue, 09-Aug-2022 06:35:41 GMT; SameSite=lax'
    ],
    server: 'Apache',
    'strict-transport-security': 'max-age=63072000; includeSubdomains;, max-age=63072000; includeSubDomains;',
    'x-frame-options': 'DENY',
    'referrer-policy': 'no-referrer-when-downgrade',
    'x-content-type-options': 'nosniff',
    'content-security-policy': "default-src 'none'; frame-ancestors 'none'",
    'x-xss-protection': '1; mode=block',
    'www-authenticate': 'Bearer realm="application"'
  },
  statusCode: 401,
  statusMessage: 'Unauthorized'
}
```

### LINKED ISSUE #142: Endless loop when user is not authenticated
**Describe the bug**
When the access token has expired we end up in an endless loop when trying to view an asset

**To Reproduce**
1. Open on localhost
2. Wait until the session has expired (I don't know how long this takes? 😬 )
3. Click on an asset

**Expected behavior**
The system should handle expired tokens. Ideally by a revalidate token, but at least not in an endless 401 loop

**Screenshots**
Skjermbilde 2022-08-12 kl  09 08 25

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**
This might be fixed if we use another auth method

### LINKED ISSUE #153: Replace MSAL with Next Auth
**In order to properly utilize Next.js' server side features, we need authn/authz to happen server side**

`msal-react`, which we are currently using, only runs client side. That blocks us from taking advantage of Next.js' server side features. By replacing `msal-react` with `next-auth`, we'll be able to move user authentication from the client, to the server. By doing that, the server will get access to perform API calls to Collibra, instead of them going from the client, through our `/api` routes, and onward to Collibra.

Resolves #129, resolves #125, resolves #124, resolves #142 
Closes #139

### PR DESCRIPTION
**Pull request by aronhoyer**
Resolves #125, resolves #153, resolves #142

### CODE REVIEW

File: components/AuthContainer/AuthContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { Container } from "../Container"
+import { Page } from "../Page"
+
+export const AuthContainer: FunctionComponent<PropsWithChildren> = ({ children }) => {
+  const { data, status } = useSession()
+
+  if (status === "loading") {
+    return (
+      <Container>
+        <Typography>Signing you in&hellip;</Typography>
```
Comment (wenche) [inline components/AuthContainer/AuthContainer.tsx:14]: Use a text snippet here?

File: components/AuthContainer/AuthContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      </Container>
+    )
+  }
+
+  if (status !== "authenticated" || data?.error === "RefreshAccessTokenError") {
+    signIn("azure-ad")
+
+    return (
+      <Container>
+        <Typography>Please sign in</Typography>
```
Comment (wenche) [inline components/AuthContainer/AuthContainer.tsx:24]: Same

File: pages/api/auth/[...nextauth].ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
+import NextAuth from "next-auth"
+import { JWT } from "next-auth/jwt"
+import ADProvider from "next-auth/providers/azure-ad"
+import qs from "query-string"
+
+import { config } from "../../../config"
```
Comment (wenche) [inline pages/api/auth/[...nextauth].ts:6]: Base url, I miss you 😢

File: pages/api/auth/[...nextauth].ts [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  expires_in: number,
+  ext_expires_in: number,
+  access_token: string,
+  refresh_token: string,
+  id_token: string
+}
+
+const tryAccessTokenRefresh = async (token: Token): Promise<Token> => {
+  try {
+    const res = await HttpClient.post<RefreshTokenResponse>(`https://login.microsoftonline.com/${config.AUTH_TENANT_ID}/oauth2/v2.0/token`, {
```
Comment (wenche) [inline pages/api/auth/[...nextauth].ts:31]: Question: Should we avoid hard coded urls like this and rather add them to some config? 🤔

**Review by wenche — CHANGES_REQUESTED** [review]
Impressive work 👏 

Even if it's sad to see so much old code go, it is satisfying as well. 

Just some small comments. As discussed on Slack, I want to just merge this code, test it for some time in the application and then adjust if needed

File: components/AuthContainer/AuthContainer.tsx (hunk 1 above)
Comment (aronhoyer) [reply components/AuthContainer/AuthContainer.tsx:14]: oh right! will fix :+1:

File: pages/api/auth/[...nextauth].ts (hunk 4 above)
Comment (aronhoyer) [reply pages/api/auth/[...nextauth].ts:31]: my take on this is that if it's needed multiple places in the code, then yes. what's your take here?

File: pages/api/auth/[...nextauth].ts (hunk 3 above)
Comment (aronhoyer) [reply pages/api/auth/[...nextauth].ts:6]: feel free to open a pr to set up the config 😅

File: pages/api/auth/[...nextauth].ts (hunk 4 above)
Comment (wenche) [reply pages/api/auth/[...nextauth].ts:31]: Jeg er pragmatisk her, men reagerer instinktivt når jeg ser url-er som tekst i kode bare 😅  Syns det er helt OK og ha det slik så lenge

**Comment by aronhoyer** [comment]
> Even if it's sad to see so much old code go

i love seeing old code go 😍

File: pages/api/auth/[...nextauth].ts (hunk 3 above)
Comment (wenche) [reply pages/api/auth/[...nextauth].ts:6]: We should try it someday, maybe it Just Works(™)

**Review by wenche — APPROVED** [review]

**Review by ShikhaMishra9 — COMMENTED** [review]
Can we consider moving duplicate {like checking  request method and  adding token to request } auth logic to one common place. May be if we can use next middleware not sure completely just read about it in next topics.

https://blog.tericcabrel.com/protect-your-api-routes-in-next-js-with-middleware/

**Comment by aronhoyer** [comment]
> like checking request method

that's generally done per request handler, but we can definitely move the token verification to a middleware :+1:

**Comment by aronhoyer** [comment]
for now, because of a weird interaction between Next Auth and middlewares, we're not going to do anything more at this point. but i've created #178 to address this in a later sprint.

**Pull request merged**